### ISSUE TITLE
Build Error: Does the ember-cli-babel version 7 and above provide no support for babel compact option?

### ISSUE THREAD

**Issue by Aswathprabhu** [babel/ember-cli-babel#302]
I was using ember-cli-babel version 6.16.0 and in my `ember-cli-build.js` I specified a config 
```javascript
babel: {
 compact: false
}
```
At this version during build I faced no errors, Recently I upgraded my ember-cli-babel version to `7.1.2` and faced an error during build `Invalid Option: compact is not a valid top-level option.` I presume this is thrown from babel-preset-env@7.0 and above. Can't we specify `compact` option for babel for ember-cli-babel@7.0 onwards(is it handled internally) or it is not supported? I an pretty sure babel [provides support](https://babeljs.io/docs/en/next/babel-generator#options) for this option.

**Comment by nicolo-ribaudo** [comment]
Try passing it as a generator option:
```
babel: {
 generatorOpts: { compact: false }
}
```

**Comment by Aswathprabhu** [comment]
@nicolo-ribaudo It fails again with `[BABEL] unknown: Invalid Option: generatorOpts is not a valid top-level option`. I presume the valid top level options are [these](https://github.com/babel/babel/blob/3ad4c70d3a0b391f16decf7b87a87165b0e39c54/packages/babel-preset-env/src/options.js#L3) If some invalid options are provided error is thrown [by babel-preset-env](https://github.com/babel/babel/blob/3ad4c70d3a0b391f16decf7b87a87165b0e39c54/packages/babel-preset-env/src/normalize-options.js#L22).

**Comment by Aswathprabhu** [comment]
@stefanpenner @rwjblue We have deleted invalid properties added to @babel/preset-env [here](https://github.com/babel/ember-cli-babel/blob/738080f5f4c7a3e2086c37713ef854e7934264bf/index.js#L456) Should we also include `compact`?

**Comment by rwjblue** [comment]
The reason the suggestion from @nicolo-ribaudo above doesn't work here is that ember-cli-babel passes the options from the `babel` config key directly in to `@babel/preset-env` and there isn't currently a way to disambiguate options destined for `@babel/preset-env` and top level configuration (e.g. `parserOpts`/`generatorOpts`).

**Comment by Aswathprabhu** [comment]
@rwjblue So do you mean that options other than the one destined for `@babel/preset-env` will not work with `ember-cli-babel`

**Comment by rwjblue** [comment]
> So do you mean that options other than the one destined for @babel/preset-env will not work with ember-cli-babel

Correct. That is the current implementation. I think we can fix this in general, but that is where things are at the moment.

**Comment by LuisAverhoff** [comment]
What is the progress on this issue? Not being able to specify `compact: false` in development is a slight issue for me as my build time has increased. This is because some of the node module like lodash and sinon once installed have sizes greater than 500KB and will emit a warning saying `styles have been deoptimized`.

**Comment by theseyi** [comment]
Have this same issue as well, build times are unacceptably longer because this option is no longer available for babel. Is there a workaround that can be implemented in the interim?

**Comment by abhishek97** [comment]
+1, build times are ~5mins due to firebase

**Comment by lvegerano** [comment]
Any updates on this we could greatly benefit from this as well

**Comment by MuseofMoose** [comment]
Pinging on this as we could also make use of it and we don't seem to be the only ones.

**Comment by yads** [comment]
Near as I can tell the deoptimization warnings that people are likely seeing are happening after this plugin and are due to ember-auto-imports and its webpack configuration. I currently have a fork of this repo that allows specifying a `babelConfig` setting on the root level `babel` configuration. This setting is sent directly to the Babel transpiler, but it only winds up affecting ember source and your application/addon code.

You can try it at this branch: https://github.com/yads/ember-cli-babel/tree/fix-for-302 and see if it helps anybody, I can look at submitting a PR if this is something that would be worthwhile.

Effectively you would do:

```
var app = new EmberApp(defaults, {
  ...
  babel: {
    babelConfig: { generatorOpts: { compact: false } }
  },
  ...
});
```

I am still trying to figure out if ember-auto-imports can be configured to allow for improving the build times.

**Comment by yads** [comment]
Ok so I think I found the issue in ember-auto-importer. The default rule for the babel loader is [here](https://github.com/ef4/ember-auto-import/blob/61922075d161aea76e8cf92340aced92e0d70510/packages/ember-auto-import/ts/webpack.ts#L150) if we could inject options into the `options` parameter there it could fix the deoptimized style warning, however, near as I can tell there's no way currently to update a config setting that is in an array, since the config merging will always concatenate arrays.

**Comment by rwjblue** [comment]
FWIW, you can specify any arbitrary Babel configuration now in the beta release (if you use a `babel.config.js`).

Doesn't resolve the ember-auto-import angle, but does unblock access to all other (previously hidden) Babel configuration options.

**Comment by yads** [comment]
Yes I saw that code path. Near as I can understand though you have to recreate the entire config that this plug-in would generate automatically. There is an issue where slow build times are acknowledged for large projects https://github.com/ef4/ember-auto-import/issues/270

**Comment by rwjblue** [comment]
> Near as I can understand though you have to recreate the entire config that this plug-in would generate automatically

Hmm, not sure what you mean. I would expect folks to be able to end up with a Babel config like mentioned in our README. 

Something like:

```js
// babel.config.js

const { buildEmberPlugins } = require('ember-cli-babel');

module.exports = function (api) {
  api.cache(true);

  return {
    presets: [
      [
        require.resolve('@babel/preset-env'),
        {
          targets: require('./config/targets'),
        },
      ],
    ],
    plugins: [...buildEmberPlugins(__dirname)],
  };
};
```

Most of the plugins are still handled automatically in that `...buildEmberPlugins(__dirname)` invocation.

**Comment by yads** [comment]
Ah that makes a lot more sense, yes this seems pretty easy to use.

**Comment by rwjblue** [comment]
I'm going to go ahead and close this, as it is now possible. Future work will happen to make this the default experience, but the fundamental feature is unblocked...

**Comment by justinzelinsky** [comment]
Hi there -- I wanted to continue the discussion here because I'm still not sure we're able to pass babel configuration keys other than `presets` and `plugins` if we write our own `babel.config.js`.

In the source, [`getBabelOptions`](https://github.com/babel/ember-cli-babel/blob/master/lib/get-babel-options.js) seems to be the function which looks for the configuration for babel but it appears to only check for `presets` and `plugins`. If I wanted to pass another property (e.g. `compact`), how does that propagate accordingly?

**Comment by rwjblue** [comment]
Hmm, I don't think so. `getBabelOptions` should not be used if you opt-in to using `babel.config.js` setup. 

See this branch:

https://github.com/babel/ember-cli-babel/blob/4c3b9091d7c711ecb804a52226409b409a702d82/index.js#L67-L85

It directly uses Babel API's to load the configuration from that file, it should support 100% of what Babel supports.

**Comment by justinzelinsky** [comment]
You are right!

Interesting -- I'm explicitly passing `compact: false` for my configuration but I'm still seeing a warning around deoptimised styling.

Thank you for the quick response.

**Comment by yads** [comment]
@justinzelinsky see my [comment](https://github.com/babel/ember-cli-babel/issues/302#issuecomment-774482717) that warning is actually coming from ember-auto-import and has no good way to override it.

**Comment by justinzelinsky** [comment]
Understood! Thanks for the help!

**Comment by mdcarreira** [comment]
Hey guys, I'm also trying to use this configuration to set `compact: false`, but I'm having difficulties in getting it right.

I'm using `ember-cli: 3.15.2` and `ember-cli-babel: 7.26.6`. I'm setting the options as explained in `ember-cli-babel` README. However, if I use the original `babel.config.js` as it is in README, I get a lot of errors like these:
```
ERROR in ./node_modules/@sentry/browser/esm/index.js
Module not found: Error: Can't resolve 'node_modules/@sentry/browser/esm/exports' in '/home/micael/code/weldnote-frontend/node_modules/@sentry/browser/esm'
 @ ./node_modules/@sentry/browser/esm/index.js 1:0-34:2
 @ /tmp/broccoli-96956DVwtAwZ2klu2/cache-891-bundler/staging/app.js
 @ multi /tmp/broccoli-96956DVwtAwZ2klu2/cache-891-bundler/staging/l.js /tmp/broccoli-96956DVwtAwZ2klu2/cache-891-bundler/staging/app.js

ERROR in ./node_modules/@sentry/browser/esm/index.js
Module not found: Error: Can't resolve 'node_modules/@sentry/browser/esm/integrations' in '/home/micael/code/weldnote-frontend/node_modules/@sentry/browser/esm'
 @ ./node_modules/@sentry/browser/esm/index.js 1:0-34:2
 @ /tmp/broccoli-96956DVwtAwZ2klu2/cache-891-bundler/staging/app.js
 @ multi /tmp/broccoli-96956DVwtAwZ2klu2/cache-891-bundler/staging/l.js /tmp/broccoli-96956DVwtAwZ2klu2/cache-891-bundler/staging/app.js

(...)
 ```

I noticed that there is an `disableModuleResolution` option and indeed setting it to `true` will allow for the server to run. But accessing the page, it crashes with the following JS error:
```
Uncaught Error: Could not find module `ember-classic-decorator` imported from `weldnote/routes/application`
    at missingModule (vendor.js:259)
    at findModule (vendor.js:270)
    at Module.findDeps (vendor.js:180)
    at findModule (vendor.js:274)
    at requireModule (vendor.js:36)
    at Class._extractDefaultExport (vendor.js:288927)
    at Class.resolveOther (vendor.js:288587)
    at Class.resolve (vendor.js:288650)
    at resolve (vendor.js:25607)
    at Registry.resolve (vendor.js:25147)
```

I tried debugging `get-babel-options.js` to see if there were any differences between the default configuration (no external file) and providing the `babel.config.js` file. I found that `ember-classic-decorator` was missing from the `plugins` list in the later solution. I tried to add it like this:

```
// full `babel.config.js` i'm using

const { buildEmberPlugins } = require('ember-cli-babel');

module.exports = function (api) {
  api.cache(true);

  return {
    presets: [
      [
        require.resolve('@babel/preset-env'),
        {
          targets: require('./config/targets'),
        },
      ],
    ],

    plugins: [
      // if you want external helpers
      [
        require.resolve('@babel/plugin-transform-runtime'),
        {
          version: require('@babel/plugin-transform-runtime/package').version,
          regenerator: false,
          useESModules: true,
        },
      ],
      ...buildEmberPlugins(__dirname, {
        disableModuleResolution: true,
      }),
      'ember-classic-decorator/lib/classic-decorator-transform-v4.js',
    ],
  };
};

```

But to no avail, the returned `options` in `get-babel-options.js` is exactly the same as without my edit. Moreover, nothing I do in `plugins` seems to affect what ends up in returned `options`. Maybe I'm looking in the wrong place?

Can somebody provide me some guidance on this? Or better yet: is there an easier way to tell `ember-cli-babel` to use auto-generated configuration file and then just change the `compact` property?

Thanks for any help!

**Comment by mdcarreira** [comment]
If you came across this post because you have big files that are being compacted and you need to see the source (as was my case), check out the solution that worked for me [in this stack overflow question](https://stackoverflow.com/questions/68740217/babel-deoptimises-styling-of-big-files-in-ember-js-app/).
